File: contador-de-palabras/contar_frecuencias.ts
const texto: string = "hola mundo hola javascript mundo mundo";

// TODO:  Divide el texto en palabras usando split()
const words: string [] = texto.split(" ");
// TODO: Crea un objeto para contar la frecuencia de cada palabra
let frecuency: Record<string, number> = {};
for (const word of words){
    if (word in frecuency){
        frecuency[word] += 1;
    } else {
        frecuency[word] = 1;
    }
}
console.log(frecuency)

// TODO: Encuentra la palabra más frecuente
const mostFrecuency: [string, number] = Object.entries(frecuency).reduce((a,b) => a[1] > b[1] ? a : b);
console.log(mostFrecuency[0])

// TODO: Filtra y muestra palabras que aparecen más de 1 vez
  const oneMore: string[] = Object.entries(frecuency).filter((count) => count[1] > 1).map((word) => word[0]);
console.log(oneMore);
  //Apuntes
  /*
Pieza del código,El ¿Por qué?,El ¿Para qué?
Object.entries(frecuency),Los objetos no son iterables por defecto como las listas.,"Para ""aplanar"" tu mapa de frecuencias en una lista de pares [palabra, cuenta] que podamos procesar."

.filter((n) => n[1] > 1),"Necesitamos descartar el ""ruido"" (palabras únicas).",Para actuar como un embudo que solo deja pasar los elementos donde el segundo valor (la frecuencia) cumple la condición.

.map((n) => n[0]),"El filtro te devuelve el par completo, pero tú solo quieres el texto.","Para ""limpiar"" el resultado, extrayendo únicamente la palabra y descartando el número que ya no necesitamos."

: string[],TypeScript necesita saber qué forma tiene el resultado final.,"Para asegurarte de que, si intentas usar oneMore después, el editor sepa que es una lista de textos y te ayude con el autocompletado."
*/

File: contador-de-palabras/contar_frecuencias.dart
import 'dart:math';

void main() {
  String texto = "hola mundo hola dart mundo mundo";

  // TODO:  Divide el texto en palabras
  List<String> words = texto.split(" "); 
  print(words);
  // TODO: Crea un mapa para contar la frecuencia de cada palabra
  Map<String, int> frecuency = {}; 
  for (String word in words) { 
    if (frecuency.containsKey(word)) { 
      frecuency[word] = frecuency[word]! + 1; 
    } else {
      frecuency[word] = 1; 
    }
  }
  // TODO:  Encuentra la palabra más frecuente
  MapEntry mostFrequentWord = frecuency.entries.reduce( 
    (a, b) => a.value > b.value ? a : b,
  );
  String winner = mostFrequentWord.key.toString();
  print(winner);
  // TODO: Imprime palabras que aparecen más de 1 vez
  List oneMore = frecuency.entries.where((n) => n.value > 1).map((n) => n.key).toList();
  print(oneMore);
}
// Apuntes
/*
Método / Propiedad,¿Por qué se usa?,¿Para qué sirve?
.split() ✂️,Para segmentar una cadena de texto larga.,Convierte un String en una List<String> basada en un separador (como un espacio).

.containsKey() 🔍,"Para evitar errores de ""valor nulo"".",Verifica si una clave ya existe en el Map antes de intentar operar con su valor.

.entries 🎟️,Para tratar el mapa como una lista de pares.,"Transforma el Map en un Iterable de objetos MapEntry, permitiendo usar métodos de listas."

.reduce() 🔄,Para consolidar toda una lista en un solo resultado.,"Compara los elementos uno a uno (como un torneo) hasta que solo queda un ""ganador""."

.where() 🧪,Para limpiar o segmentar datos.,Filtra la colección y deja pasar solo los elementos que cumplen una condición lógica.

.map() 🎨,Para transformar la naturaleza de los datos.,Toma cada elemento y lo convierte en algo distinto (ej: de un objeto MapEntry a solo un String).

.toList() 📦,"Para ""materializar"" el resultado.",Convierte los procesos intermedios (Iterables) en una lista real que puedes guardar o imprimir.
*/